File: desktop-mend/src/main/java/co/samco/mend4/desktop/commands/EncryptFromStdIn.java
package co.samco.mend4.desktop.commands;

import co.samco.mend4.core.OSDao;
import co.samco.mend4.core.Settings;
import co.samco.mend4.core.exception.CorruptSettingsException;
import co.samco.mend4.desktop.core.I18N;
import co.samco.mend4.desktop.helper.CryptoHelper;
import co.samco.mend4.desktop.helper.FileResolveHelper;
import co.samco.mend4.desktop.helper.InputHelper;
import co.samco.mend4.desktop.output.PrintStreamProvider;

import javax.crypto.BadPaddingException;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import javax.inject.Inject;
import java.io.IOException;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;
import java.util.function.Function;

public class EncryptFromStdIn extends Encrypt {
    public static final String COMMAND_NAME = "enci";

    private OSDao osDao;

    private final List<Function<List<String>, List<String>>> behaviourChain = Arrays.asList(
            a -> assertSettingsPresent(a),
            a -> shouldDropHeader(a),
            a -> encryptFromInput(a)
    );

    @Inject
    public EncryptFromStdIn(Settings settings, PrintStreamProvider log, I18N strings, CryptoHelper cryptoHelper,
                            InputHelper inputHelper, OSDao osDao, FileResolveHelper fileResolveHelper) {
        super(settings, log, strings, cryptoHelper, inputHelper, fileResolveHelper);
        this.osDao = osDao;
    }

    private List<String> encryptFromInput(List<String> args) {
        Scanner scanner = new Scanner(osDao.getStdIn());
        StringBuilder sb = new StringBuilder();

        while (scanner.hasNextLine()) {
            sb.append(scanner.nextLine());
            if (scanner.hasNextLine()) {
                sb.append(strings.getNewLine());
            }
        }
        scanner.close();

        try {
            cryptoHelper.encryptTextToLog(sb.toString().toCharArray(), dropHeader);
        } catch (IOException | CorruptSettingsException | InvalidKeySpecException | NoSuchAlgorithmException
                | IllegalBlockSizeException | InvalidKeyException | BadPaddingException
                | InvalidAlgorithmParameterException | NoSuchPaddingException e) {
            e.printStackTrace();
        }
        return null;
    }

    @Override
    public void execute(List<String> args) {
        executeBehaviourChain(behaviourChain, args);
    }

    @Override
    public List<String> getCommandAliases() {
        return Arrays.asList(COMMAND_NAME);
    }

    @Override
    public String getUsageText() {
        return strings.getf("EncryptFromStdIn.usage", COMMAND_NAME, APPEND_FLAG);
    }

    @Override
    public String getDescriptionText() {
        return strings.get("EncryptFromStdIn.description");
    }
}

File: desktop-mend/src/main/java/co/samco/mend4/desktop/commands/EncryptFromStdIn.kt
package co.samco.mend4.desktop.commands

import co.samco.mend4.core.Settings
import co.samco.mend4.desktop.core.I18N
import co.samco.mend4.desktop.helper.CryptoHelper
import co.samco.mend4.desktop.helper.FileResolveHelper
import co.samco.mend4.desktop.helper.InputHelper
import co.samco.mend4.desktop.helper.SettingsHelper
import co.samco.mend4.desktop.output.PrintStreamProvider
import java.util.Scanner
import java.util.function.Function
import javax.inject.Inject

class EncryptFromStdIn @Inject constructor(
    settings: Settings,
    settingsHelper: SettingsHelper,
    log: PrintStreamProvider,
    strings: I18N,
    cryptoHelper: CryptoHelper,
    inputHelper: InputHelper,
    fileResolveHelper: FileResolveHelper
) : Encrypt(
    settings = settings,
    settingsHelper = settingsHelper,
    log = log,
    strings = strings,
    cryptoHelper = cryptoHelper,
    inputHelper = inputHelper,
    fileResolveHelper = fileResolveHelper
) {
    private val behaviourChain = listOf(
        Function { a: List<String> -> assertSettingsPresent(a) },
        Function { a: List<String> -> shouldDropHeader(a) },
        Function { a: List<String> -> encryptFromInput(a) }
    )

    private fun encryptFromInput(args: List<String>): List<String>? {
        val scanner = Scanner(System.`in`)
        val sb = StringBuilder()
        while (scanner.hasNextLine()) {
            sb.append(scanner.nextLine())
            if (scanner.hasNextLine()) {
                sb.append(strings.newLine)
            }
        }
        scanner.close()
        try {
            cryptoHelper.encryptTextToLog(sb.toString().toCharArray(), dropHeader)
        } catch (t: Throwable) {
            failWithMessage(log, t.message)
        }
        return null
    }

    override fun execute(args: List<String>) {
        executeBehaviourChain(behaviourChain, args)
    }

    override val commandAliases: List<String>
        get() = listOf(COMMAND_NAME)
    override val usageText: String
        get() = strings.getf("EncryptFromStdIn.usage", COMMAND_NAME, APPEND_FLAG)
    override val descriptionText: String
        get() = strings["EncryptFromStdIn.description"]

    companion object {
        const val COMMAND_NAME = "enci"
    }
}